### PR TITLE
Add setting to enforce full replication of the cluster before deleting Pods

### PR DESCRIPTION
**Pull request by johscheuer**
# Description

This is the "dumb" version of https://github.com/FoundationDB/fdb-kubernetes-operator/issues/901 without any additional checks (e.g. are we targeting a stateless process). I decided to use `false` as the default since that reflects the current behaviour and the user can/has decide if that's a valid risk or not.

## Type of change

- New feature (non-breaking change which adds functionality)

# Discussion

I decided to query the FDB cluster for the latest status, one could also argue that the "cached" status from the FDB cluster object should be good enough.

# Testing

Unit tests.

# Documentation

I added a few lines about that feature but I think it's self describing.

# Follow-up

https://github.com/FoundationDB/fdb-kubernetes-operator/issues/901

### CODE REVIEW

File: controllers/remove_process_groups_test.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+	fdbtypes "github.com/FoundationDB/fdb-kubernetes-operator/api/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+
+	"k8s.io/utils/pointer"
+)
+
+var _ = Describe("remove_process_groups", func() {
+	var cluster *fdbtypes.FoundationDBCluster
+	//var fakeConnectionString string
```
Comment (johscheuer) [inline controllers/remove_process_groups_test.go:39]: TODO remove these lines

File: controllers/remove_process_groups_test.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+	BeforeEach(func() {
+		cluster = internal.CreateDefaultCluster()
+		//fakeConnectionString = "operator-test:asdfasf@127.0.0.1:4501"
+	})
+
+	Describe("Reconciliation", func() {
```
Comment (johscheuer) [inline controllers/remove_process_groups_test.go:46]: Remove the describe block

File: controllers/remove_process_groups.go [hunk 3]
```diff
@@ -51,6 +52,25 @@ func (u RemoveProcessGroups) Reconcile(r *FoundationDBClusterReconciler, context
 		return nil
 	}
 
+	// We don't use the "cached" of the cluster status from the CRD to minimize the window between data loss (e.g. a node
+	// or a set of Pods is not reachable anymore). We still end up with the risk to actually query the FDB cluster and after that
+	// query the cluster gets into a degraded state.
+	// We could be smarter here and only block removals that target stateful processes by e.g. filtering hose out of the
```
Comment (johscheuer) [inline controllers/remove_process_groups.go:58]:
```suggestion
	// We could be smarter here and only block removals that target stateful processes by e.g. filtering those out of the
```

**Review by ltsampros — APPROVED** [review]
I like the PR but happy to discuss the following points:
1) `EnforceFullReplicationForDeletion` should be true by default
2) Do we need the knob in the first place, if we make this behaviour the default one?

**Comment by johscheuer** [comment]
> I like the PR but happy to discuss the following points:
> 
> 1. `EnforceFullReplicationForDeletion` should be true by default

I decided to default to `false` to not change the current behaviour. Changing this setting is more or less a noop from the cluster point of view. Since we are still in `0.*` we could argue that this is fine. Let's chat about that offline and decide what's the best way to go forward.

> 2. Do we need the knob in the first place, if we make this behaviour the default one?

I think it can be useful for a user to decide if they require these stricter checks or not. Also it's useful to be able to disable this check if something is not working as expected or the cluster is in a weird state and disabling this check would actually help (I don't have a specific case here in mind).

**Review by brownleej — APPROVED** [review]
👍

**Comment by johscheuer** [comment]
Given the offline discussions and the feedback in the related issue I make the default `true` but still leave the setting here for user to enable/disable the setting.

**Pull request merged**